src/core/jit/transform/input_resolver.rs

```diff
@@ -50,63 +51,71 @@ where
     <Output as TryFrom<serde_json::Value>>::Error: std::fmt::Debug,
 {
     pub fn resolve_input(
-        &self,
+        self,
         variables: &Variables<Output>,
     ) -> Result<OperationPlan<Output>, ResolveInputError> {
-        let new_fields = self
+        let index = self.plan.index;
+        let selection = self
             .plan
             .selection
-            .iter()
-            .map(|field| (*field).clone().try_map(&|value| value.resolve(variables)))
-            .map(|field| match field {
-                Ok(field) => {
-                    let field = self.iter_args(field)?;
-                    Ok(field)
-                }
-                Err(err) => Err(err),
-            })
+            .into_iter()
+            .map(|field| field.try_map(&|value| value.resolve(variables)))
             .collect::<Result<Vec<_>, _>>()?;
-        Ok(OperationPlan::new(
-            self.plan.root_name(),
-            new_fields,
-            self.plan.operation_type(),
-            self.plan.index.clone(),
-            self.plan.is_introspection_query,
-        ))
+
+        // Traverse over fields again to verify/populate defaults for args
+        // because the previous iteration will just try convert values based on
+        // variables ignoring default values in schema and not checking if arg
+        // is required TODO: consider changing [Field::try_map] to be able to do
+        // this check?
+        let selection = selection
+            .into_iter()
+            .map(|field| Self::resolve_field(&index, field))
+            .collect::<Result<_, _>>()?;
```

Comment:
```suggestion
        let index = self.plan.index;
        let selection = self
            .plan
            .selection
            .into_iter()
            .map(|field| field.try_map(&|value| value.resolve(variables)))
            // Traverse over fields again to verify/populate defaults for args
            // because the previous iteration will just try convert values based on
            // variables ignoring default values in schema and not checking if arg
            // is required.
            .map(|field| Self::resolve_field(&index, field?))
            .collect::<Result<Vec<_>, _>>()?;
```

Reply:
thanks, applied